js/src/blocks/product-select-with-text-field/edit.js

```diff
@@ -0,0 +1,81 @@
+/**
+ * External dependencies
+ */
+import { useWooBlockProps } from '@woocommerce/block-templates';
+import { SelectControl } from '@wordpress/components';
+import { useState } from '@wordpress/element';
+import {
+	__experimentalUseProductEntityProp as useProductEntityProp,
+	__experimentalTextControl as TextControl,
+} from '@woocommerce/product-editor';
+
+/**
+ * Internal dependencies
+ */
+import { Label } from '../components';
+
+const FALLBACK_VALUE = '';
+
+export default function Edit( { attributes, context } ) {
+	const { options, customInputValue } = attributes;
+
+	const blockProps = useWooBlockProps( attributes );
+	const [ value, setValue ] = useProductEntityProp( attributes.property, {
+		postType: context.postType,
+		fallbackValue: FALLBACK_VALUE,
+	} );
+
+	// Refer to the "Derived value for initialization" in README for why this `useState`
+	// can initialize directly.
+	const [ optionValue, setOptionValue ] = useState( () => {
+		// Empty string is a valid option value.
+		const initValue = value ?? FALLBACK_VALUE;
+		const selectedOption = options.find(
+			( option ) => option.value === initValue
+		);
+
+		return selectedOption?.value ?? customInputValue;
+	} );
+
+	const isSelectedCustomInput = optionValue === customInputValue;
+
+	const [ text, setText ] = useState( isSelectedCustomInput ? value : '' );
+
+	const handleSelectionChange = ( nextOptionValue ) => {
+		setOptionValue( nextOptionValue );
+
+		if ( nextOptionValue === customInputValue ) {
+			setValue( text );
+		} else {
+			setValue( nextOptionValue );
+		}
```

Comment:
💅  This can be reduced to:

`setValue( nextOptionValue === customInputValue ? text : nextOptionValue );`

Reply:
I lead toward keeping it as it is.